### PR TITLE
FEATURE: Added end date to UserCommons and Commons, cow health and cow milking jobs only run before end date/after start date

### PR DESCRIPTION
**Pull request by epuzio**
**Details:**
Based on Professor Conrad's ["As an admin, I can control dates for game play"](https://github.com/ucsb-cs156/proj-happycows/issues/42#issue-1738786926) feature. 

**Description:**
Before this PR, each commons had a start date, but no end date, and the start date didn't affect game play. In this PR, I added an end date to Commons and a start and end date to userCommons. I also added a gameInProgress() function to the userCommons entity that checks against LocalDateTime.now() and tested this function. Next, I added Last Date as a field on the AdminCreateCommonsPage/AdminEditCommonsPage form. Finally, I added calls to gameInProgress on MilkTheCowsJob and UpdateCowHealthJob to skip over any userCommons whose game is not currently in progress. Our Instructor Reports Job is a stub, so I did not implement any changes there.

**Implementation of Professor Conrad's ideas:**
- [x] Add lastday date. (this is more clear than end date, since it indicates that the game ends at 11:59:59 on that day). Needs to be added to entity in backend, POST and PUT endpoints in backend, form in frontend, and adjusted on Create and Update pages.
- [x]  Add a function to the UserCommons entity (and tests for it) called gameInProgress() that returns true if today's date is >= start date, and <= end date. This could be it's own PR, with tests. Note that mocking "today's date" is possible; tricky, but possible.
- [x]  Add logic to the backend of the milk the cows job, where it iterates over the commons. Skip over commons where gameInProgress() returns false.
- [x]  Do the same for the updateCowHealth job.
- [ ]  Do the same for instructor reports job

**Screenshots:**
Currently: Screenshot 2023-06-06 at 2 31 04 PM

Screenshot 2023-06-06 at 12 09 39 PM

Screenshot 2023-06-06 at 12 10 14 PM

Screenshot 2023-06-06 at 1 03 27 PM

Screenshot 2023-06-06 at 1 04 01 PM

Storybook:
Previously: https://ucsb-cs156-s23.github.io/proj-happycows-s23-6pm-3/storybook/?path=/story/components-commons-commonsform--uncontrolled
Currently: https://ucsb-cs156-s23.github.io/proj-happycows-s23-6pm-3/prs/59/storybook/?path=/story/components-commons-commonsform--uncontrolled

### CODE REVIEW

File: frontend/src/main/components/Commons/CommonsForm.js
Comment (MrJacobEisner) [inline frontend/src/main/components/Commons/CommonsForm.js:1]: Do you think the default value should be today?

**Review by MrJacobEisner — DISMISSED** [review]

File: frontend/src/main/components/Commons/CommonsForm.js
Comment (epuzio) [reply frontend/src/main/components/Commons/CommonsForm.js:1]: For lastDate? Good point, I can fix that.

**Review by MaximillianStewart — DISMISSED** [review]
LGTM

**Review by awaang — DISMISSED** [review]

File: package.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  "main": "index.js",
+  "directories": {
+    "doc": "docs"
+  },
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "ISC"
+}
```
Comment (pconrad) [inline package.json]: Why is this here?   My guess is that someone did an npm install in the wrong directory and created this by accident.   

We only want a package.json inside the frontend directory.

Can we remove this from the PR?

File: src/main/java/edu/ucsb/cs156/happiercows/jobs/MilkTheCowsJob.java [hunk 2]
```diff
@@ -34,23 +35,27 @@ public String formatDollars(double amount) {
...
+                Iterable<UserCommons> allUserCommons = userCommonsRepository.findByCommonsId(commons.getId());
 
-            for (UserCommons userCommons : allUserCommons) {
-                milkCows(ctx, commons, userCommons);
-            }
-        }
-
-        ctx.log("Cows have been milked!");
+                for (UserCommons userCommons : allUserCommons) {
+                    if(userCommons.gameInProgress()){
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/happiercows/jobs/MilkTheCowsJob.java]: Why are we checking this at the userCommons level?

The game is in progress, or not, at the Commons level, not at the UserCommons level.

It doesn't make sense to iterate over every player and make the same check for each one. This check should be at the outer loop.

File: src/main/java/edu/ucsb/cs156/happiercows/entities/UserCommons.java [hunk 3]
```diff
@@ -33,5 +33,14 @@ public class UserCommons {
   private int numOfCows;
 
   private double cowHealth;
+
+  private LocalDateTime startingDate;
+
+  private LocalDateTime lastDate;
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/happiercows/entities/UserCommons.java]: Why?   These fields go on the Commons, not the UserCommons!

File: src/main/java/edu/ucsb/cs156/happiercows/entities/Commons.java [hunk 4]
```diff
@@ -29,6 +29,7 @@
   private double milkPrice;
   private double startingBalance;
   private LocalDateTime startingDate;
+  private LocalDateTime lastDate;
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/happiercows/entities/Commons.java:31]: You have the fields here.  You do not need them, and should not store them redundantly on the UserCommons.

File: src/main/java/edu/ucsb/cs156/happiercows/entities/UserCommons.java [hunk 5]
```diff
@@ -33,5 +33,14 @@ public class UserCommons {
...
+
+  public boolean gameInProgress(){
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/happiercows/entities/UserCommons.java]: Please move this function to the Commons.  It belongs there, not on the UserCommons.

File: src/main/java/edu/ucsb/cs156/happiercows/jobs/MilkTheCowsJob.java [hunk 6]
```diff
@@ -34,23 +35,27 @@ public String formatDollars(double amount) {
...
+            Iterable<Commons> allCommons = commonsRepository.findAll();
 
-        for (Commons commons : allCommons) {
-            String name = commons.getName();
-            double milkPrice = commons.getMilkPrice();
-            ctx.log("Milking cows for Commons: " + name + ", Milk Price: " + formatDollars(milkPrice));
+            for (Commons commons : allCommons) {
+                String name = commons.getName();
+                double milkPrice = commons.getMilkPrice();
+                ctx.log("Milking cows for Commons: " + name + ", Milk Price: " + formatDollars(milkPrice));
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/happiercows/jobs/MilkTheCowsJob.java]: This is where you should check:

```suggestion
                if (!commons.gameInProgress()) {
                    ctx.log("Commons " + name + " is not currently in progress, and cows were not milked.");
                } else { 
                   ctx.log("Milking cows for Commons: " + name + ", Milk Price: " + formatDollars(milkPrice));
                   // etc...  be sure to close this brace...
```

File: src/main/java/edu/ucsb/cs156/happiercows/jobs/UpdateCowHealthJob.java [hunk 7]
```diff
@@ -46,22 +46,26 @@ public void accept(JobContext ctx) throws Exception {
                 User user = userRepository.findById(userCommons.getUserId()).orElseThrow(()->new RuntimeException("Error calling userRepository.findById(" + userCommons.getUserId() + ")"));
                 ctx.log("User: " + user.getFullName() + ", numCows: " + userCommons.getNumOfCows() + ", cowHealth: " + userCommons.getCowHealth());
 
-                double newCowHealth = calculateNewCowHealth(userCommons.getCowHealth(), userCommons.getNumOfCows(), totalCows, carryingCapacity, degradationRate);
-                ctx.log(" old cow health: " + userCommons.getCowHealth() + ", new cow health: " + newCowHealth);
-                userCommons.setCowHealth(newCowHealth);
+                if(userCommons.gameInProgress()){
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/happiercows/jobs/UpdateCowHealthJob.java]: Please fix this one in a similar way

File: src/main/java/edu/ucsb/cs156/happiercows/jobs/UpdateCowHealthJob.java [hunk 8]
```diff
@@ -46,22 +46,26 @@ public void accept(JobContext ctx) throws Exception {
...
-            double oldCowHealth,
-            int numCows,
-            int totalCows,
-            int carryingCapacity,
-            double degradationRate) {
+        double oldCowHealth,
+        int numCows,
+        int totalCows,
+        int carryingCapacity,
+        double degradationRate, boolean gameInProgress) {
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/happiercows/jobs/UpdateCowHealthJob.java]: Why do we need this?

We can just avoid calling the function in the first place.  Note that you didn't use the value inside the function.

**Review by pconrad — CHANGES_REQUESTED** [review]
A few things... mostly having to do with NOT storing redundant data on the UserCommons when we can get it through the commons.

**Review by MrJacobEisner — APPROVED** [review]
Some whitespace issues, but the code looks good. Aside from that,  assuming Professor Conrad's comment is resolved, LGTM.

**Review by awaang — APPROVED** [review]
LGTM

**Review by pconrad — APPROVED** [review]
LGTM!  Will test on staff-qa before merging

**Pull request merged**